### DISCUSSION TITLE
Copy a dll to AppX folder and Publish Folder

### DISCUSSION THREAD

**Discussion by ahipki** [dotnet/maui#23628] [Q&A]
Hello,

I need a dll : vcinet.x64.dll.
When I work in local environment, I always need to copy this dll to the folder AppX because the folder doesn't want to be copied in AppX folder. The dll is in the parent folder. If I had understand AppX utility, it's only for when I run the app.
So I decided to put this dll in my project files and a script to copy it.
I put the dll on Resources\vcinet\x64\vcinet.x64.dll

I put this on my csproj : 
```
<ItemGroup>
  <Content Include="Resources\vcinet\x64\vcinet.x64.dll">
    <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    <CopyToPublishDirectory>Always</CopyToPublishDirectory>
  </Content>
</ItemGroup>
```

And this : 
```
 <ItemGroup>
   <None Update="Resources\vcinet\x64\vcinet.x64.dll">
     <CopyToOutputDirectory>Always</CopyToOutputDirectory>
   </None>
 </ItemGroup>
```

And I want to add a "script" : 
I tried : 
```
  <Target Name="CopyVcinetDllToOutputDirectory" AfterTargets="Build">
    <Copy SourceFiles="Resources\vcinet\x64\vcinet.x64.dll" DestinationFolder="$(OutputPath)annie" />
  </Target>
```
 It **works**, I have my annie folder created with the dll when I build.
When I publish : 
```
  <Target Name="CopyVcinetDllToPublishDirectory" AfterTargets="Publish">
    <Copy SourceFiles="Resources\vcinet\x64\vcinet.x64.dll" DestinationFolder="$(PublishDir)vcinet\x64" />
  </Target
```
If I use 7-zip to unpackage the .msix, I don't see any vicinet folder
I tried $(TargetDir) instead of $(PublishDir) and also $(OutDir) and also $(OutputPath) and it doesn't work.
I also tried : 
```
  <Target Name="CopyVcinetDllToPublishDirectory" AfterTargets="Publish">
    <Exec Command="xcopy /Y /I /E &quot;$(TargetDir)Resources\vcinet\x64\vcinet.x64.dll&quot; &quot;$(PublishDir)vcinet\x64\vcinet.x64.dll&quot;" />
  </Target>
```
I tried with AfterPublish instead of Publish and it doesn't work
I also tried this but not working either.
```
  <Target Name="CopyVcinetDllToPublishDirectory" BeforeTargets="_ConvertItems;_CalculateInputsForGenerateCurrentProjectAppxManifest">
    <Copy sourcefiles="Resources\vcinet\x64\vcinet.x64.dll" DestinationFolder="$(PublishDir)vcinet\x64" />
  </Target>
```
---------------

I also want, for the debug, the dll in my AppX folder so I tried 
```
  <Target Name="CopyVcinetDllToOutputDirectory" AfterTargets="Build">
    <Copy SourceFiles="Resources\vcinet\x64\vcinet.x64.dll" DestinationFolder="$(OutputPath)AppX\vcinet\x64" />
  </Target>
```

I see my vcinet folder in AppX folder.
But when I launch the app, the AppX folder disappear and reappear without de folder vcinet

Thanks if you find me the miracle command :)

(To have the file in my AppX, I moved from Resources to vcinet directly at the root of the project)

**Comment by GMudide** [comment]
any suggestions. I'm stumble upon same problem.

**Comment by siprix** [comment]
This works for me:
```
<ItemGroup>    	    
    <MauiAsset   Include=Resources\vcinet\x64\vcinet.x64.dll" LogicalName="vcinet.x64.dll" />
</ItemGroup>
```

**Reply by robreno** [reply]
Did the making the DLL a Maui asset solve the problem?

**Reply by siprix** [reply]
Yes, works good.
It could be also checked in the file '[app].build.appxrecipe', created during the build.
Added asset causes to generate element 'AppxPackagedFile' with 'Include=' which matches <MauiAsset Include=> 'PackagePath' which matches 'LogicalName'.

Applied here: https://github.com/siprix/SampleMaui

**Reply by robreno** [reply]
Thanks, that is good information. I am using F# projects in a Maui solution to provide Lexing/Parsing of a DSL for a query engine. I need to integrate FsLexYacc into the solution so when I create the MSIX it bundles the Package libraries required to support FsLexYacc. This might be just what I need.
